_posts/2022-01-21-safe-releases.md

```diff
@@ -0,0 +1,95 @@
+
+Below I outline a general process which can also be applied to multi-system releases, while noting that there will generally be many more releases involved in order to gracefully transition each system safely while it is still interacting with other apps.
+
+### Does your release contain long-running data migrations?
+
+One common request in a software business is to release a feature and prefill a bunch of data. Since Rails migrations support full Ruby and Rails code, it's tempting to just write a loop in there and populate a bunch of records.
+
+This is dangerous! Without knowing exactly how long a migration can take, you run the risk of timing out your release commands and failing the build. Depending on your configuration and whether you moved this outside of a transaction, this could leave you with an inconsistent DB state.
+
+### Does your release add indices, locks, or perform other heavy table operations?
+
+It's also important to be aware of expensive database operations like adding locks or indices, changing column types, etc.
+
+There are great resources available [1] to learn whether a particular action is safe or not, and as you learn more about them you will also learn strategies to release these changes without any downtime or user-facing side-effects.
+
+## 1-2-3: A slow, safe release process
+
+For the next piece, we'll take a simple scenario and describe the releases necessary for a safe migration. The request is: move the `email` column from the `user` table to the `user_details` table. Keep in mind that there are existing emails in the original table.
+
+My rule of thumb is: The more explicit, smaller steps, the safer the release (to a point!). The more you break things down into parts, the less likely any release will result in a broken build or put your application into a messy state that requires manual intervention.
+
+Using that rule, here's a well-organized plan to release this change:
+
+### Release 1 - Prepare your new column, begin data synchronization, and asynchronously prepare all data
+
+First and foremost, you'll want to create your new `email` column on the `user_details` table. If you were to run only this, you would have no data in the new column and therefore would not be able to use it yet with your application code.
+
+To ensure data is fully synched by the time the second release comes along, you'll want two things:
+Modify any code that previously set `email` on the `user` table to also update the `user_details` table. By updating both, you will make sure that any updates after your population scripts are reflected in the second table.
+An asynchronous task to backpopulate emails in the new table. Based on our additional considerations, we should not be populating large amounts of data in the migration itself. So a typical approach here would be to ship a rake task or other script that can be run on the server once the release goes out.
```

Comment:
"An asynchronous task to backpopulate emails in the new table" - Is this a bullet-point? The sentence seems incomplete.

Reply:
Yep, was incomplete/bad formatting